File: 1293-shortest-path-in-a-grid-with-obstacles-elimination/1293-shortest-path-in-a-grid-with-obstacles-elimination.java
class Solution {
    int[][] directions = new int[][]{{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    
    public int shortestPath(int[][] grid, int k) {
        int m = grid.length, n = grid[0].length;
        boolean[][][] visited = new boolean[m][n][k+1];
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{0, 0, k});
        int steps = 0;
        while(!queue.isEmpty()){
            int size = queue.size();
            for(int i = 0; i < size; ++i){
                int[]curr = queue.poll();
                if(curr[0] == m - 1 && curr[1] == n - 1) return steps;
                
                for(int[] dir : directions){
                    int row = curr[0] + dir[0], col = curr[1] + dir[1], elims = curr[2];
                    if(row >= 0 && row < m && col >= 0 && col < n){
                        elims -= grid[row][col];
                        if(elims >= 0 && !visited[row][col][elims]){
                            visited[row][col][elims] = true;
                            queue.offer(new int[]{row, col, elims});
                        }
                    }
                }
            }
            ++steps;
        }
        
        return -1;
    }
}

File: 1293-shortest-path-in-a-grid-with-obstacles-elimination/1293-shortest-path-in-a-grid-with-obstacles-elimination.py
from collections import deque
class Solution:
    def shortestPath(self, grid: List[List[int]], k: int) -> int:
        queue = deque()
        visited = set()
        m = len(grid)
        n = len(grid[0])
        
        directions = [[-1, 0], [1, 0], [0, -1], [0, 1]]
        queue.append((0, 0, k))
        visited.add((0, 0, k))
        steps = 0
        while queue:
            size = len(queue)
            for i in range(size):
                curr = queue.popleft()
                if curr[0] == m - 1 and curr[1] == n - 1:
                    return steps
                
                for dir in directions:
                    row = curr[0] + dir[0]
                    col = curr[1] + dir[1]
                    curr_k = curr[2]
                    if row >= 0 and row < m and col >= 0 and col < n:
                        curr_k -= grid[row][col]
                        candidate = (row, col, curr_k)
                        if curr_k >= 0 and candidate not in visited:
                            visited.add(candidate)
                            queue.append(candidate)
        
            steps += 1
        return -1
                
        